File: src/algorithms/trees/properties/binary-tree-tilt/BinaryTreeTilt_test.java
// javac *.java && java -ea BinaryTreeTilt_test
public class BinaryTreeTilt_test {
    static TiltNode makeNode(int value, TiltNode left, TiltNode right) {
        TiltNode node = new TiltNode(value);
        node.left = left;
        node.right = right;
        return node;
    }

    static TiltNode leaf(int value) { return new TiltNode(value); }

    public static void main(String[] args) {
        BinaryTreeTilt algo = new BinaryTreeTilt();

        // test: null root returns 0
        assert algo.binaryTreeTilt(null) == 0 : "Null root should return 0";

        // test: single node returns 0
        assert algo.binaryTreeTilt(leaf(1)) == 0 : "Single node should return 0";

        // test: simple 3-node tree
        assert algo.binaryTreeTilt(makeNode(1, leaf(2), leaf(3))) == 1 : "3-node tilt should be 1";

        // test: non-negative for 7-node tree
        TiltNode tree = makeNode(4, makeNode(2, leaf(1), leaf(3)), makeNode(6, leaf(5), leaf(7)));
        assert algo.binaryTreeTilt(tree) >= 0 : "Tilt should be non-negative";

        System.out.println("All tests passed!");
    }
}

File: src/algorithms/trees/properties/binary-tree-tilt/BinaryTreeTilt_test.cpp
// g++ -o btt_test BinaryTreeTilt_test.cpp && ./btt_test
#include "sources/BinaryTreeTilt.cpp"
#include <cassert>
#include <iostream>

TreeNode* makeBTTNode(int value, TreeNode* left = nullptr, TreeNode* right = nullptr) {
    TreeNode* node = new TreeNode(value);
    node->left = left;
    node->right = right;
    return node;
}

int main() {
    // test: null root returns 0
    assert(binaryTreeTilt(nullptr) == 0);

    // test: single node returns 0
    assert(binaryTreeTilt(makeBTTNode(1)) == 0);

    // test: simple 3-node tree
    assert(binaryTreeTilt(makeBTTNode(1, makeBTTNode(2), makeBTTNode(3))) == 1);

    // test: non-negative for 7-node tree
    TreeNode* tree = makeBTTNode(4,
        makeBTTNode(2, makeBTTNode(1), makeBTTNode(3)),
        makeBTTNode(6, makeBTTNode(5), makeBTTNode(7)));
    assert(binaryTreeTilt(tree) >= 0);

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
